server/src/main/java/org/opensearch/index/seqno/LocalCheckpointTracker.java

```diff
@@ -134,6 +134,29 @@ public synchronized void markSeqNoAsPersisted(final long seqNo) {
         markSeqNo(seqNo, persistedCheckpoint, persistedSeqNo);
     }
 
+    /**
+     * Updates the processed checkpoint with the provided sequence number if segment replication is enabled
+     *
+     * @param seqNo the sequence number to mark as processed
+     */
+    public synchronized void segrepMarkSeqNoAsProcessed(final long seqNo) {
+        markSegrepSeqNo(seqNo, processedCheckpoint, persistedCheckpoint);
+    }
+
+    @SuppressForbidden(reason = "Object#notifyAll")
+    private void markSegrepSeqNo(final long seqNo, final AtomicLong processedCheckpoint, final AtomicLong persistedCheckpoint) {
+        assert Thread.holdsLock(this);
+        advanceMaxSeqNo(seqNo);
+        if ((seqNo > persistedCheckpoint.get()) || (seqNo <= processedCheckpoint.get())) {
+            return;
+        }
+        try {
+            processedCheckpoint.compareAndSet(processedCheckpoint.get(), seqNo);
```

Comment:
Shouldn't this be using `updateCheckpoint` to ensure that the `persistedSeqNo` is kept up to date?

Reply:
The call to updateCheckpoint to ensure persistedSeqNo is up to date still takes place when markSeqNoAsPersisted is called. In the case of segrep, we update processedSeqNo just once at the end since indexing doesn't take place on the replica - we omit the call to updateCheckpoint since it checks that all seq numbers till that point are consecutively processed.

Reply:
Thanks for the explanation! I think we should capture this bit in the Javadoc for `fastForwardProcessedSeqNo`:

> In the case of segrep, we update processedSeqNo just once at the end since indexing doesn't take place on the replica - we omit the call to updateCheckpoint since it checks that all seq numbers till that point are consecutively processed.